keel-sqs/src/main/kotlin/com/netflix/spinnaker/keel/sqs/SqsResourceCheckListener.kt

```diff
@@ -27,38 +34,48 @@ internal class SqsResourceCheckListener(
   private val objectMapper: ObjectMapper,
   private val actuator: ResourceActuator
 ) {
-  private lateinit var listenerJob: Job
+  private lateinit var rootJob: Job
 
-  private val scope: CoroutineScope
-    get() = CoroutineScope(Default)
+  private val scope: CoroutineScope = CoroutineScope(IO)
 
   @PostConstruct
   fun startListening() {
-    listenerJob = scope.launch {
-      while (isActive) {
-        handleMessages()
+    rootJob = scope.launch {
+      val channel = Channel<Message>()
+      launchMessageReceiver(channel)
+      repeat(10) {
+        launchWorker(channel)
       }
     }
   }
 
-  internal fun handleMessages() {
-    with(sqsClient) {
-      receiveMessages()
+  private fun CoroutineScope.launchMessageReceiver(channel: SendChannel<Message>) = launch {
+    runUntilCancelled {
+      sqsClient
+        .receiveMessages()
         .also {
           if (it.isNotEmpty()) log.debug("Got {} messages: {}", it.size, it.map(Message::getMessageId))
         }
         .forEach {
-          try {
-            with(it.parse()) {
-              actuator.checkResource(name.let(::ResourceName), apiVersion, kind)
-            }
-            deleteMessage(it)
-          } catch (e: JsonProcessingException) {
-            log.error("Could not parse message payload: {}", it.body)
-          } catch (e: Throwable) {
-            log.error("Resource Actuator threw {}: {}", e.javaClass.simpleName, e.message)
+          channel.send(it)
+        }
+    }
+  }
+
+  private fun CoroutineScope.launchWorker(channel: ReceiveChannel<Message>) = launch {
+    runUntilCancelled {
+      for (message in channel) {
+        try {
```

Comment:
I wonder if its worth doing a call to SQS to set the visibilityTimeout here before starting to work on the message to ensure that it is still a valid message handle and that you have time remaining to work on it

I don't know if there is a case where your channel gets backed up if all the workers are busy and the message takes too long to dispatch to a worker

Probably pretty theoretical for the scale at which this queue would be operating

Reply:
also if we don't care about two different consumers processing the same message than this probably doesn't matter either

Reply:
Yeah, the operations are currently idempotent in theory. The SQS queue just lets work propagate across the cluster. I still need to get the scheduler to be smarter because currently all the instances are going to schedule checks for all the things!